lib: add AsSSortedList method ...

ssiccha opened the pull request:
... for collections that already are strictly sorted lists.

This ensures that `AsSet([1..n])` returns `[1..n]` instead of creating a
new plain list.

ChrisJefferson commented:
Stupid question -- why does this have to be "IsCollection and IsSSortedList", and not just "IsSSortedList"?

ssiccha commented:
Good question. Since there is a method installed for `IsCollection and IsConstantTimeAccessList` I was afraid that `RankFilter` may be to small then, but I didn't check. :) Actually we have
```
gap> RankFilter(IsSSortedList);
8
gap> RankFilter(IsCollection and IsConstantTimeAccessList);
6
```

I have updated the PR. I have added a "manual rank increment" such that the new functions should always be called when the argument already is a (immutable or mutable) `IsSSortedList`.

fingolfin commented on lib/list.gi:
```diff
@@ -646,6 +646,18 @@ InstallOtherMethod(AsSSortedList,
+InstallMethod( AsSSortedList,
+    "for a mutable strictly sorted list",
+    [ IsSSortedList and IsMutable ],
+    100
```
Please don't use static rank adjustments like that except as a very, very last resort.

In this particular case, it would probably be fine to just use `SUM_FLAGS`. Alternatively, one could use a dynamic  rank adjustment via a rank function, which has the added benefit that if done right, it is "self-documenting", i.e., makes it clear which other filter you are trying to beat. But I think `SUM_FLAGS` is fine here.

ssiccha replied:
Will do! Thanks.

ssiccha replied:
Done. I also updated the commit message accordingly.

fingolfin commented on lib/list.gi:
```diff
@@ -646,6 +646,18 @@ InstallOtherMethod(AsSSortedList,
+InstallMethod( AsSSortedList,
+    "for a strictly sorted list",
+    [ IsSSortedList ],
+    100
```
Ditto.

ssiccha commented:
I'll have a look into why tests fail.

ChrisJefferson commented on lib/list.gi:
```diff
@@ -646,6 +646,18 @@ InstallOtherMethod(AsSSortedList,
+    SUM_FLAGS,
+    Immutable );
+
+InstallMethod( AsSSortedList,
```
I understand why we might have this optimisation (ID_FUNC), but Immutable is fast (and doesn't copy) objects which are already immutable, and I never like functions which require another function to be correct.

ChrisJefferson replied:
Also, maybe chuck in some more tests (I'd be tempted to chuck in a sorted list of booleans, and a sorted string) as they are "non-standard" list types. Probably not required in this case, but I personally use them whenever I'm sorting lists.

ChrisJefferson commented:
Thanks
